server/src/main/java/org/elasticsearch/rest/MethodHandlers.java

```diff
@@ -56,11 +59,29 @@ MethodHandlers addMethods(RestHandler handler, RestRequest.Method... methods) {
     }
 
     /**
-     * Returns the handler for the given method or {@code null} if none exists.
+     * Return a handler for a given method and a version
+     * If a handler for a given version is not found, the handler for Version.CURRENT is returned.
+     * We only expect Version.CURRENT or Version.CURRENT -1. This is validated.
+     *
+     * Handlers can be registered under the same path and method, but will require to have different versions (CURRENT or CURRENT-1)
+     *
+     * //todo What if a handler was added in V8 but was not present in V7?
+     *
+     * @param method a REST method under which a handler was registered
+     * @param version a Version under which a handler was registered
+     * @return a handler
      */
     @Nullable
-    RestHandler getHandler(RestRequest.Method method) {
-        return methodHandlers.get(method);
+    RestHandler getHandler(RestRequest.Method method, Version version) {
+        Map<Version, RestHandler> versionToHandlers = methodHandlers.get(method);
+
+        if(versionToHandlers == null){
+            return null;
+        }
+        if (versionToHandlers.containsKey(version)) {
```

Comment:
Is it ever valid for an entry in the map to have a null handler? I would think not so this could be simplified to:

```
final RestHandler handler = versionToHandlers.get(version);
return handler != null || version.equals(Version.CURRENT) ? handler : versionToHandlers.get(Version.CURRENT);
```

Reply:
agree - I don't expect null values in `versionToHandlers` map. Will apply the suggestion